### ISSUE TITLE
LFX Workspace: A stream log processing framework for WasmEdge

### ISSUE THREAD

**Issue by markcty** [WasmEdge/WasmEdge#2552]
## Motivation

For motivation, plz read [feat: A stream log processing framework for WasmEdge · Issue #2470 · WasmEdge/WasmEdge (github.com)](https://github.com/WasmEdge/WasmEdge/issues/2470).

The goal of this project is to provide a framework that can be used to WebAssembly applications that collect, transform, and route logs. It should consist of the following components:

* Collector: The collectors are responsible for collecting logs from various services (e.g., MySQL Binlog, Docker, Kubernetes, and File).
* Transformer: Transformers manipulate logs emitted from collectors and other transformers. They can be chained together to form a pipeline (DAG).
* Dispatcher: forward the processed logs to their final destinations (e.g., Kafka, Redis, ElasticSearch)
* Router: configure the behavior of the framework and run all components as a whole

### Milestones

- [x] [Week 1] (Jun. 1 - Jun.11) Project setup and initial design
  * Research popular log processing framework like [vector](https://vector.dev/) and [filebeat](https://www.elastic.co/beats/filebeat)
  * Draft the project architecture and relavent traits (`Collector`, `Dispatcher`, `Router`, and `Transformer`)
  * Write a dummy implementation for these traits to show how the framework should be used
- [x] [Week 2 - 3] (Jun. 12 - Jun. 25) Verify the feasibility of the initial design
  * Implement a real-world example combination of `Collector`, `Dispatcher`, `Router`, and `Transformer` (No need for full functionalities, but should be complete enough for the framework to run as a whole)
- [x] [Week 4] (Jun. 26 - Jul. 2) Refine and finalize the design
  * Present the initial design to the community and gather feedbacks
  * Refine and finalize the design for the framework
- [x] [Week 5 - 8] (Jul. 3 - Jul. 30) Implement these triats
  - Implement at least two `Collector`s. One for MySQL database binlog and the other for a generic log file in a Linux container in the same Kubernetes pod
  - Implement at least two `Transformer` algorithms supported by [FileBeat](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-overview.html)
  - Implement at least three `Dispatcher`s. One for a Kafka queue, one for a Redis database, and the other for ElasticSearch
  - Implement configurable `Router`s for simple routing rules for the above combos of `Collector`s and `Dispatcher`s
  - Provide CI and test code
  - Provide benchmark
- [x] [Week 9] (Jul. 31 - Aug. 13) Initial exposure
  * Write docs and tutorials
  * Present the initial version to the communitiy and gather feedbacks
- [x] [Week 10 - 11] (Aug. 14 - Aug. 20) Project polishing and hardening
- [x] [Week 12] (Aug. 21 - Aug. 31) Spare time

### Repository

https://github.com/markcty/wasm-log-flex

**Comment by markcty** [comment]
## W1 Update

TL;DR

* Temporarily added MySQL8.0 support for `mysql-binlog-kafka` demo (use my `mysql_native_password` auth plugin).
* Researched popular log processing frameworks such as `vector`,`logstash` and `maxwells`. 
* Drafted initial design for our framework and write some demo code.

### Design

Like many other popular frameworks, our log processing framework will consist of three main `component`s:

* `Collector`: collecting logs from various services (e.g., MySQL Binlog, Docker, Kubernetes, and File).
* `Transformer`: manipulate logs emitted from collectors or other transformers.
* `Dispatcher`: dispatch processed logs to their final destinations (e.g., Kafka, Redis, ElasticSearch).

Components communicate with each other using custom `Event`. They can be chained together to form a pipeline.

Here is a demo combination of various components.

![Architecture drawio-6214831](https://github.com/WasmEdge/WasmEdge/assets/54015375/7d0ab630-1f70-4eb2-a15d-6208b2b402fa)


### Implementation

See this branch for the initial implementation [[markcty/wasm-log-flex at draft-design (github.com)](https://github.com/markcty/wasm-log-flex/tree/draft-design)](https://github.com/markcty/wasm-log-flex/tree/draft-design).

Each component should have a unique id which is used to route the events. API of `Component`:

```Rust
pub trait ComponentApi: 'static + Send + Sync {
    fn id(&self) -> &str;
    fn kind(&self) -> ComponentKind;
}
pub enum ComponentKind {
    Collector,
    Transformer,
    Dispatcher,
}
```

All the components use an `event_hub` to transmit events. The `event_hub` concept is borrowed from the hub in computer network. All components should be first connected to the hub before events can be sent to each other. API of `EventHub`:

```Rust
pub enum Event {
    BinlogEvent(BinlogEvent),
    BinlogWithTable(BinlogEventParsed),
}
#[async_trait]
pub trait EventHubApi {
    async fn send_event(&self, event: Event, component_id: &str) -> Result<(), Error>;
    async fn poll_event(&self, component_id: &str) -> Result<Event, Error>;
    fn register_component(&mut self, collector: &impl ComponentApi);
}
```

Here is a demo on how this frame should be used:

```Rust
// create an event hub
let mut hub = EventHub::new();

// create a collector, a transformer, and a dispatcher first
// register them in the `EventHub`
let collector = BinlogCollector;
let transformer = BinlogTableParser;
let dispatcher = KafkaMysqlTableDispatcher;
hub.register_component(&collector);
hub.register_component(&transformer);
hub.register_component(&dispatcher);

// start all the components
let hub = Arc::new(hub);
let hub_arc = Arc::clone(&hub);
tokio::spawn(async move {
    collector.start_collecting(hub_arc).await;
});
let hub_arc = Arc::clone(&hub);
tokio::spawn(async move {
    transformer.start_parsing(hub_arc).await;
});
let hub_arc = Arc::clone(&hub);
tokio::spawn(async move {
    dispatcher.start_dispatching(hub_arc).await;
});
```

### Q&A about the implementation 

If you have any questions regarding the implementation, comment below and I will update my answer in this section.

> How should event be represented? Dynamic or static disptach?

Because of Rust's strict type system requirement, it's hard to make a perfect definition for the `Event`. Some of the solutions are:

##### I. `Event` as an enum(current solution)

Pros:

* Customizable: any type can fit into event
* Performance: static type is fast

Cons:

* Not so convenient when a new event type needs to be added

##### II. `Event` as a trait(like `Box<dyn Event>`)

Pros:

* Flexible: developers can define their event in their own crate, thus allowing them to create collectors, transformers, and dispatchers out of this crate.

Cons:

* Dynamic dispatch is expensive
* Downcast code is hard to write sometimes

##### III. `Event` as an Json-like struct that contains nested `HashMap`

Pros:

* Compromise between flexibility and customizability
* Performance is good

Cons:

* Not so expressive as a real type

Currently, I am using the first option. I think I will go with the third option in futher development.

**Comment by juntao** [comment]
Sorry I thought I responded to this already. But apparently I have not!

Yes, using the enum types is good for now. But there might be many different dispatchers, and each dispatcher might require additional config (eg to dispatch logs associated with a specific table to one or more designated Kafka topics). So, I guess we should at least explore option #3 for dispatchers.

Also, maybe we can call it EventRouter instead of EventHub?

**Comment by markcty** [comment]
> Sorry I thought I responded to this already. But apparently I have not!
> 
> Yes, using the enum types is good for now. But there might be many different dispatchers, and each dispatcher might require additional config (eg to dispatch logs associated with a specific table to one or more designated Kafka topics). So, I guess we should at least explore option #3 for dispatchers.
> 
> Also, maybe we can call it EventRouter instead of EventHub?
> 
> 

Thank you for your response and advice! I will rename the EventHub to EventRouter and try option3. More updates is coming soon...

**Comment by markcty** [comment]
## W2-3 Update

### TL;DR

* Fix `mysql_cdc`(the binlog event replicator)'s crash when connecting MySQL8. The bug actually came from the `mysql_common` package. I've [created an PR for it and got merged](https://github.com/blackbeam/rust_mysql_common/pull/94).
* Make `mysql_cdc` async. Previously, `mysql_cdc` uses sync io which could block the thread. This makes it hard to run in wasm environment since it's a single thread runtime(I guess?).
* Re-designed `Event` mentioned in the last update.
* Implement a mysql binlog collector, a binlog event filter, and a Kafka dispatcher. They can know run together as a whole.

### Demo of work

See [wasm-log-flex/demos/src/bin/mysql_table_kafka.rs at main · markcty/wasm-log-flex · GitHub](https://github.com/markcty/wasm-log-flex/blob/main/demos/src/bin/mysql_table_kafka.rs)

Now, we have a complete demo of how the framework and relavent components should be used.

Firstly, user creates a binlog collector by specifying mysql configuration.

```Rust
let options = ReplicaOptions {
    username: String::from("root"),
    password: String::from("password"),
    blocking: true,
    ssl_mode: SslMode::Disabled,
    binlog: BinlogOptions::from_end(),
    ..Default::default()
};
let collector = BinlogCollector::new("binlog_collector", "binlog_parser", options);
```

Secondly, user creates a binlog filter by giving it a set of rules. Current available rules are to include or exclude events from the specified database or table.

```Rust
let rules = BinlogFilterRules::new()
    .exclude("d1", None) // exclude all events from database 'd1'
    .include("d1", Some("t1")); // include events from database 'd1' table 't1'
let transformer = BinlogFilter::new("binlog_parser", "kafka_dispatcher", rules);
```

Thirdly, user creates a Kafka dispatcher. The user can also specify the topic these binlog events will be sent to. In the topic string, `%{database}` and `%{table}` will be replaced by the real one. In this demo, events from database `d1` table `t1` will be sent to topic `logFlex.d1.t1`.

```rust
let mut dispatcher =
    KafkaDispatcher::new("kafka_dispatcher", vec!["127.0.0.1:9092".to_string()]);
dispatcher.set_topic(r"logFlex.%{database}.%{table}");
```

Finally, these components are registered in the `EventRouter` and get started.

### Implementation

#### Binlog Collector

The binlog collector has roughly the same funtionality with the one in the pretest. I've put much effort into adding async support and fixing crash problem when connecting MySQL8 this 2 weeks.

The `Event` format generated by the binlog collector are largely borrowed from the [maxwell's deamon](https://maxwells-daemon.io/dataformat/).

#### Binlog Filter

`BinlogFilters` are configured by a set of rules(`BinlogFilterRules`). Rules are evaluted one by one.

```rust
pub struct BinlogFilterRules {
    rules: Vec<BinlogFilterRule>,
}

enum BinlogFilterRule {
    Include {
        database: String,
        table: Option<String>,
    },
    Exclude {
        database: String,
        table: Option<String>,
    },
}
```

#### Kafka Table Dispatcher

The kafka table dispatcher now supports specifying topics for different events. It will now have local cache of all the topics in kafka and create one if needed.

**Comment by juntao** [comment]
Thank you! I think it looks very good overall. We should ensure the design is flexible enough to address a diverse set of use cases.

* How do I use custom transformers that transform the data in the event stream (e.g., to change the DateTime to a Unix timestamp, or to create a JSON structure from the raw event for downstream processing)?
* Can we create an example that collects log from a file? I am thinking an NGINX or APACHE server log file. In this context, how do the {database} and {table} symbols in the dispatcher work?

**Comment by markcty** [comment]
> Thank you! I think it looks very good overall. We should ensure the design is flexible enough to address a diverse set of use cases.
> 
> * How do I use custom transformers that transform the data in the event stream (e.g., to change the DateTime to a Unix timestamp, or to create a JSON structure from the raw event for downstream processing)?
> * Can we create an example that collects log from a file? I am thinking an NGINX or APACHE server log file. In this context, how do the {database} and {table} symbols in the dispatcher work?

Thank you for your suggestions! To maximize flexibility and make serialization/deserialization easier, I think I will use `serde_json`'s `Value` instead of customized `Value` then. And for your second question, how about making a rule that the words inside the bracket should be viewed as a path into the event value. For example, `%{/sql/table}` will be replaced by `sql.table` in the json value and the user will need to make sure that `sql.table` exists in the event value.

Updates will come soon~

**Comment by juntao** [comment]
That sounds good. Thank you! Look forward to it.

You can have a look at the syntax of Maxwell's config files. Mayer our json config can have a similar approach so that Maxwell users will be able to migrate to our system quickly. 😂

**Comment by markcty** [comment]
> That sounds good. Thank you! Look forward to it.
> 
> You can have a look at the syntax of Maxwell's config files. Mayer our json config can have a similar approach so that Maxwell users will be able to migrate to our system quickly. 😂

Sorry for the late delivery. The truth is that I am blocked by some unknown bug from the tokio_wasi or the wasmedge runtime.

I am trying to build a log file collector. It requires the collector to read new new log lines when new logs arrive. Since filesystem notification library are not available in wasm, I had to use a more naive method that is to read the log file periodically. This method involves waiting in between two file readings. That's where the problem comes. The wasmedge runtime seems to be blocked and never wake up after sleeping. 

A minimal reproducible code would be:
```Rust
use std::time::Duration;

#[tokio::main(flavor = "current_thread")]
async fn main() {
    loop {
        println!("dummy");
        tokio::time::sleep(Duration::from_millis(100)).await;
    }
}
```

The code will run for a few times and stopped?

So, I am wondering if it is possible that you could give me some directions to look into the `tokio_wasi` or `wasmedge`?

**Comment by juntao** [comment]
@L-jasmine Can you look into this? It seems to be an async issue in the guest. Thanks.

**Comment by juntao** [comment]
Also, perhaps we can investigate what it takes to get filesystem notifications? If this is a native function, perhaps we can create a WasmEdge plugin for it? The plugin will receive file system notifications, and call an API function in the Wasm guest app when a notification arrives. @markcty

**Comment by markcty** [comment]
> Also, perhaps we can investigate what it takes to get filesystem notifications? If this is a native function, perhaps we can create a WasmEdge plugin for it? The plugin will receive file system notifications, and call an API function in the Wasm guest app when a notification arrives. @markcty 

Sure, I will look into it~

**Comment by markcty** [comment]
## Milestone 4[week 5-8] quick update

As milestone 4 lasts for a long period, I will give a quick update about my progress and ask for some suggestions. More details about the implementation will be posted as soon as the milestone is over.

Milestone 4 is about implementing all the collectors, transformers and dispatchers. Current status are:

Collectors:
- MySQL binlog collector: working now. log analyzing ability to be enhanced
- File collector: pending(domain socket support for wasmedge is still pending https://github.com/WasmEdge/WasmEdge/pull/2216)

Transformers:
- Binlog Event Filter: able to filter out events from specified table or database
- Event Replicator: able to replicate events and send them to different dispatchers

Dispatchers:
- Kafka Dispatcher: done
- Redis Dispatcher: done. support four kinds of operation: `xadd` to stream, `LPush` or `RPush` to stream, `Pub` to channel
- Elastic Search: to be done

In addition to these components, I've implemented a `wlf-aio`(wasm-log-flex-all-in-one) wasm binary(https://github.com/markcty/wasm-log-flex/tree/main/wlf-aio). It aims to provide an all-in-one binary for the user to use all the components easily. 

Its configuration options are largely borrowed from Maxwell'
s daemon. Below is an example config that will forward and filter logs and dispatch them to both redis and kafka:

```yaml
collectors:
  - id: binlog_collector
    type: BinlogCollector
    destination: filter
    user: root
    password: password
transformers:
  - id: filter
    type: BinlogFilter
    destination: replicator
    rules:
      - exclude:
          database: d1
      - include:
          database: d1
          table: t1
  - id: replicator
    type: EventReplicator
    destinations:
      - redis
      - kafka
dispatchers:
  - id: kafka
    type: KafkaDispatcher
    topic: logFlex.%{/meta/database}.%{/sql/table}
    bootstrap_brokers: ["127.0.0.1:9092"]
  - id: redis
    type: RedisDispatcher
    mode:
      type: Pub
      channel: logFlex.%{/meta/database}.%{/sql/table}
```

@juntao Would you plz have a look and see if it meets your expectations? And, as you have mentioned in the project specification, you would want to have two transformer algorithm implemented that are supported by Filebeat. So, can you plz recommend two algorithms for me to implement?

**Comment by juntao** [comment]
@markcty My apologies for the delay. This design looks good. I would like to stay close to the Maxwell syntax -- perhaps even providing a Python script to convert from Maxwell config to ours. :)

So, where are we now? Do you have a demo I can run? Thanks!

**Comment by markcty** [comment]
> @markcty My apologies for the delay. This design looks good. I would like to stay close to the Maxwell syntax -- perhaps even providing a Python script to convert from Maxwell config to ours. :)
> 
> So, where are we now? Do you have a demo I can run? Thanks!

Thank you for the reply. 

We now have a demo in the README(https://github.com/markcty/wasm-log-flex#quick-startusing-wlf-aio).

During the past few weeks, I have written the tutorial, added CI, add more support for Binlog Event parsing. 

I also tried to add a generic log file collector. But it is pending because of the lack of unix domain socket support in WasmEdge(https://github.com/WasmEdge/WasmEdge/pull/2216). 

In addition, I have tried to adapting the elastic search rust client to webassembly but it didn't work because multi-thread is not available in WasmEdge. I noticed that the WasmEdge community is planning to add thread support(https://github.com/WasmEdge/WasmEdge/issues/1467). Does it mean that multi-threaded tokio runtime will be supported soon? If so, developing async application like this would be much more easier😂.

I will start to work on providing the conversion script you mentioned then.

**Comment by juntao** [comment]
That is great. I will check out the examples soon.

Is it possible for us to support all config options from Maxwell? That would allow a 1-1 conversation between the two systems.

AF sockets #2216 will be supported in the next release of WasmEdge at the end of this month.

I think it is possible to adopt a tokio-based elastic search client to Rust + Wasm. What you need is to use the [tokio_wasi](https://crates.io/crates/tokio_wasi) crate instead of the standard tokio crate. The reason is that async socket is not yet a WASI standard and hence cannot be merged upstream to regular tokio. Here are some examples for using tokio_wasi to adapt existing Rust async I/O libraries:

[hyper_wasi](https://github.com/WasmEdge/hyper) | [Examples](https://github.com/WasmEdge/wasmedge_hyper_demo)
[mysql_async_wasi](https://github.com/WasmEdge/mysql_async_wasi) | [Examples](https://github.com/WasmEdge/wasmedge-db-examples/tree/main/mysql_async)

**Comment by markcty** [comment]
> That is great. I will check out the examples soon.
> 
> Is it possible for us to support all config options from Maxwell? That would allow a 1-1 conversation between the two systems.
> 
> AF sockets #2216 will be supported in the next release of WasmEdge at the end of this month.
> 
> I think it is possible to adopt a tokio-based elastic search client to Rust + Wasm. What you need is to use the [tokio_wasi](https://crates.io/crates/tokio_wasi) crate instead of the standard tokio crate. The reason is that async socket is not yet a WASI standard and hence cannot be merged upstream to regular tokio. Here are some examples for using tokio_wasi to adapt existing Rust async I/O libraries:
> 
> [hyper_wasi](https://github.com/WasmEdge/hyper) | [Examples](https://github.com/WasmEdge/wasmedge_hyper_demo) [mysql_async_wasi](https://github.com/WasmEdge/mysql_async_wasi) | [Examples](https://github.com/WasmEdge/wasmedge-db-examples/tree/main/mysql_async)

Sorry, I made a mistake about why the elastic search client won't run. I have successfully replaced tokio with tokio_wasi and hacked all the code for it to be compiled to webassembly. But it failed to run in WasmEdge and reported a really confusing error.

<img width="1103" alt="image" src="https://github.com/WasmEdge/WasmEdge/assets/54015375/f930e99d-1d27-401b-a34e-d397db01f17a">

**Comment by juntao** [comment]
Hmm, it tries to call a function that is not available. It could be a libc function that is not available in WASI. 

Can you provide a full example that we can debug? Thanks!

**Comment by markcty** [comment]
> Hmm, it tries to call a function that is not available. It could be a libc function that is not available in WASI.
> 
> Can you provide a full example that we can debug? Thanks!

Sure, I've forked it(https://github.com/markcty/elasticsearch-rs). 

1. `cargo build -p elasticsearch --target wasm32-wasi --example wasm --no-default-features`
2. `wasmedge target/wasm32-wasi/debug/examples/wasm.wasm`
3. You shall see the error.

**Comment by markcty** [comment]
Hi, I've add the support for reading maxwell configuration file to the `wlf-aio`. And I've fixed the issue for elastic client. Dispatcher for elasticsearch can also be used now.

**Comment by juntao** [comment]
This is great! I think we should publish an `elasticsearch_wasi` crate for other people to use. This is a very nice side outcome for this internship. Thank you!

**Comment by dannypsnl** [comment]
Close completed issue, feel free to reopen it if I'm wrong.
